doc/man/man3/XmForm.3
'\" t
...\" Form.sgm /main/9 1996/09/08 20:45:00 rws $
.de P!
.fl
\!!1 setgray
.fl
\\&.\"
.fl
\!!0 setgray
.fl			\" force out current output buffer
\!!save /psv exch def currentpoint translate 0 0 moveto
\!!/showpage{}def
.fl			\" prolog
.sy sed -e 's/^/!/' \\$1\" bring in postscript file
\!!psv restore
.
.de pF
.ie     \\*(f1 .ds f1 \\n(.f
.el .ie \\*(f2 .ds f2 \\n(.f
.el .ie \\*(f3 .ds f3 \\n(.f
.el .ie \\*(f4 .ds f4 \\n(.f
.el .tm ? font overflow
.ft \\$1
..
.de fP
.ie     !\\*(f4 \{\
.	ft \\*(f4
.	ds f4\"
'	br \}
.el .ie !\\*(f3 \{\
.	ft \\*(f3
.	ds f3\"
'	br \}
.el .ie !\\*(f2 \{\
.	ft \\*(f2
.	ds f2\"
'	br \}
.el .ie !\\*(f1 \{\
.	ft \\*(f1
.	ds f1\"
'	br \}
.el .tm ? font underflow
..
.ds f1\"
.ds f2\"
.ds f3\"
.ds f4\"
.ta 8n 16n 24n 32n 40n 48n 56n 64n 72n
.TH "XmForm" "library call"
.SH "NAME"
\fBXmForm\fP \(em The Form widget class
.iX "XmForm"
.iX "widget class" "Form"
.SH "SYNOPSIS"
.PP
.nf
#include <Xm/Form\&.h>
.fi
.SH "DESCRIPTION"
.PP
Form is a container widget with no input semantics of its own\&.
Constraints are placed on children of the Form to define attachments
for each of the child\&'s four sides\&.
These attachments can be to the Form, to another child widget or gadget,
to a relative position within the Form, or to the initial position of
the child\&.
The attachments determine the layout behavior of the Form when resizing
occurs\&.
.PP
The default value for \fBXmNinitialFocus\fP is the value of
\fBXmNdefaultButton\fP\&.
.PP
Following are some important considerations in using a Form:
.IP "   \(bu" 6
Every child must have an attachment on either the left or the right\&.
If initialization or \fBXtSetValues\fP leaves a widget without
such an attachment, the result depends upon the value of
\fBXmNrubberPositioning\fP\&.
.IP "" 10
If \fBXmNrubberPositioning\fP is False, the child is given an
\fBXmNleftAttachment\fP of \fBXmATTACH_FORM\fP and an
\fBXmNleftOffset\fP equal to its current \fIx\fP value\&.
.IP "" 10
If \fBXmNrubberPositioning\fP is True, the child is given an
\fBXmNleftAttachment\fP of \fBXmATTACH_POSITION\fP and an
\fBXmNleftPosition\fP proportional to the current \fIx\fP value divided
by the width of the Form\&.
.IP "" 10
In either case, if the child has not been previously given an \fIx\fP
value, its \fIx\fP value is taken to be 0 (zero), which places the child at the
left side of the Form\&.
.IP "   \(bu" 6
If you want to create a child without any attachments, and then
later (for example, after creating and managing it, but before realizing it)
give it a right attachment through \fBXtSetValues\fP, you must set its
\fBXmNleftAttachment\fP to \fBXmATTACH_NONE\fP at the same time\&.
.IP "   \(bu" 6
The \fBXmNresizable\fP resource controls only whether a geometry request
by the child will be granted\&.
It has no effect on whether the child\&'s size can be changed because of
changes in geometry of the Form or of other children\&.
.IP "   \(bu" 6
Every child has a preferred width, based on geometry requests it
makes (whether they are granted or not)\&.
.IP "   \(bu" 6
If a child has attachments on both the left and the right sides,
its size is completely controlled by the Form\&.
It can be shrunk below its preferred width or enlarged above it, if
necessary, due to other constraints\&.
In addition, the child\&'s geometry requests to change its own width may
be refused\&.
.IP "   \(bu" 6
If a child has attachments on only its left or right side, it will
always be at its preferred width (if resizable, otherwise at is current
width)\&.
This may cause it to be clipped by the Form or by other children\&.
.IP "   \(bu" 6
If a child\&'s left (or right) attachment is set to \fBXmATTACH_SELF\fP, its
corresponding left (or right) offset is forced to 0 (zero)\&.
The attachment is then changed to \fBXmATTACH_POSITION\fP, with a
position that corresponds to the \fIx\fP value of the child\&'s left (or
right) edge\&.
To fix the position of a side at a specific \fIx\fP value, use
\fBXmATTACH_FORM\fP or \fBXmATTACH_OPPOSITE_FORM\fP with the \fIx\fP
value as the left (or right) offset\&.
.IP "   \(bu" 6
Unmapping a child has no effect on the Form except that the child
is not mapped\&.
.IP "   \(bu" 6
Unmanaging a child unmaps it\&.
If no other child is attached to it, or if all children attached to it
and all children recursively attached to them are also all unmanaged,
all of those children are treated as if they did not exist in
determining the size of the Form\&.
.IP "   \(bu" 6
When using \fBXtSetValues\fP to change the \fBXmNx\fP resource of a
child, you must simultaneously set its left attachment to either
\fBXmATTACH_SELF\fP or \fBXmATTACH_NONE\fP\&.
Otherwise, the request is not granted\&.
If \fBXmNresizable\fP is False, the request is granted only if the
child\&'s size can remain the same\&.
.IP "   \(bu" 6
A left (or right) attachment of \fBXmATTACH_WIDGET\fP, where
\fBXmNleftWidget\fP (or \fBXmNrightWidget\fP) is NULL, acts like an
attachment of \fBXmATTACH_FORM\fP\&.
.IP "   \(bu" 6
If an attachment is made to a widget that is not a child of the
Form, but an ancestor of the widget is a child of the Form, the
attachment is made to the ancestor\&.
.PP
All these considerations are true of top and bottom attachments as well,
with top acting like left, bottom acting like right, \fIy\fP acting like
\fIx\fP, and height acting like width\&.
.SS "Classes"
.PP
Form inherits behavior, resources, and traits from \fBCore\fP,
\fBComposite\fP, \fBConstraint\fP,
\fBXmManager\fP, and \fBXmBulletinBoard\fP\&.
.PP
The class pointer is \fBxmFormWidgetClass\fP\&.
.PP
The class name is \fBXmForm\fP\&.
.SS "New Resources"
.PP
The following table defines a set of widget resources used by the
programmer to specify data\&. The programmer can also set the resource
values for the inherited classes to set attributes for this widget\&.
To reference a resource by name or by class in a \fB\&.Xdefaults\fP file,
remove the \fBXmN\fP or \fBXmC\fP prefix and use the remaining letters\&.
To specify one of the defined values for a resource in a \fB\&.Xdefaults\fP
file, remove the \fBXm\fP prefix and use the remaining letters (in
either lowercase or uppercase, but include any underscores between
words)\&. The codes in the access column indicate if the given resource
can be set at creation time (C), set by using \fBXtSetValues\fP
(S), retrieved by using \fBXtGetValues\fP (G), or is not
applicable (N/A)\&.
.PP
.TS
tab() box;
c s s s s
l| l| l| l| l.
\fBXmForm Resource Set\fP
\fBName\fP\fBClass\fP\fBType\fP\fBDefault\fP\fBAccess\fP
_____
XmNfractionBaseXmCMaxValueint100CSG
_____
XmNhorizontalSpacingXmCSpacingDimension0CSG
_____
XmNrubberPositioningXmCRubberPositioningBooleanFalseCSG
_____
XmNverticalSpacingXmCSpacingDimension0CSG
_____
.TE
.IP "\fBXmNfractionBase\fP" 10
Specifies the denominator used in calculating the relative position of
a child widget using \fBXmATTACH_POSITION\fP constraints\&.
The value must not be 0 (zero)\&.
.IP "" 10
If the value of a child\&'s \fBXmNleftAttachment\fP (or
\fBXmNrightAttachment\fP) is \fBXmATTACH_POSITION\fP, the position of
the left (or right) side of the child is relative to the left
side of the Form and is a fraction of the width of the Form\&.
This fraction is the value of the child\&'s \fBXmNleftPosition\fP (or
\fBXmNrightPosition\fP) resource divided by the value of the Form\&'s
\fBXmNfractionBase\fP\&.
.IP "" 10
If the value of a child\&'s \fBXmNtopAttachment\fP (or
\fBXmNbottomAttachment\fP) is \fBXmATTACH_POSITION\fP, the position of
the top (or bottom) side of the child is relative to the top
side
of the Form and is a fraction of the height of the Form\&.
This fraction is the value of the child\&'s \fBXmNtopPosition\fP (or
\fBXmNbottomPosition\fP) resource divided by the value of the Form\&'s
\fBXmNfractionBase\fP\&.
.IP "\fBXmNhorizontalSpacing\fP" 10
Specifies the offset for right and left attachments\&. This resource is
only used if no offset resource is specified (when attaching to a
widget), or if no margin resource is specified (when attaching
to the Form)\&.
.IP "\fBXmNrubberPositioning\fP" 10
Indicates the default near (left) and top attachments for a child of the
Form\&.
(Note that whether this resource actually applies to the left or
right side of the child and its attachment
depends on the value of
the \fBXmNlayoutDirection\fP\&.)
.IP "" 10
The default left attachment is applied whenever initialization or
\fBXtSetValues\fP leaves the child without either a left or right
attachment\&.
The default top attachment is applied whenever initialization or
\fBXtSetValues\fP leaves the child without either a top or bottom
attachment\&.
.IP "" 10
If this Boolean resource is set to False, \fBXmNleftAttachment\fP and
\fBXmNtopAttachment\fP default to \fBXmATTACH_FORM\fP,
\fBXmNleftOffset\fP defaults to the current \fIx\fP value of the left
side of the child,
and \fBXmNtopOffset\fP defaults to the current
\fIy\fP value of the child\&.
The effect is to position the child according to its absolute distance
from the left or top side of the Form\&.
.IP "" 10
If this resource is set to True, \fBXmNleftAttachment\fP and
\fBXmNtopAttachment\fP default to \fBXmATTACH_POSITION\fP,
\fBXmNleftPosition\fP defaults to a value proportional to the current
\fIx\fP value of the left side of the child divided by the width of the
Form, and \fBXmNtopPosition\fP defaults to a value proportional to the
current \fIy\fP value of the child divided by the height of the Form\&.
The effect is to position the child relative to the left or top
side of the Form and in proportion to the width or height of the Form\&.
.IP "\fBXmNverticalSpacing\fP" 10
Specifies the offset for top and bottom attachments\&. This resource is
only used if no offset resource is specified (when attaching to a
widget), or if no margin resource is specified (when attaching
to the Form)\&.
.IP "" 10
.TS
tab() box;
c s s s s
l| l| l| l| l.
\fBXmForm Constraint Resource Set\fP
\fBName\fP\fBClass\fP\fBType\fP\fBDefault\fP\fBAccess\fP
_____
XmNbottomAttachmentXmCAttachmentunsigned charXmATTACH_NONECSG
_____
XmNbottomOffsetXmCOffsetint0CSG
_____
XmNbottomPositionXmCPositionint0CSG
_____
XmNbottomWidgetXmCWidgetWidgetNULLCSG
_____
XmNleftAttachmentXmCAttachmentunsigned charXmATTACH_NONECSG
_____
XmNleftOffsetXmCOffsetint0CSG
_____
XmNleftPositionXmCPositionint0CSG
_____
XmNleftWidgetXmCWidgetWidgetNULLCSG
_____
XmNresizableXmCBooleanBooleanTrueCSG
_____
XmNrightAttachmentXmCAttachmentunsigned charXmATTACH_NONECSG
_____
XmNrightOffsetXmCOffsetint0CSG
_____
XmNrightPositionXmCPositionint0CSG
_____
XmNrightWidgetXmCWidgetWidgetNULLCSG
_____
XmNtopAttachmentXmCAttachmentunsigned charXmATTACH_NONECSG
_____
XmNtopOffsetXmCOffsetint0CSG
_____
XmNtopPositionXmCPositionint0CSG
_____
XmNtopWidgetXmCWidgetWidgetNULLCSG
_____
.TE
.IP "\fBXmNbottomAttachment\fP" 10
Specifies attachment of the bottom side of the child\&. It can have the
following values:
.RS
.IP "\fBXmATTACH_NONE\fP" 10
Do not attach the bottom side of the child\&.
.IP "\fBXmATTACH_FORM\fP" 10
Attach the bottom side of the child to the bottom
side of the Form\&.
.IP "\fBXmATTACH_OPPOSITE_FORM\fP" 10
Attach the bottom side of the child to
the top side of the Form\&.
\fBXmNbottomOffset\fP can be used to determine the visibility of the
child\&.
.IP "\fBXmATTACH_WIDGET\fP" 10
Attach the bottom side of the child to the top
side of the widget or
gadget specified in the \fBXmNbottomWidget\fP resource\&.
If \fBXmNbottomWidget\fP is NULL, \fBXmATTACH_WIDGET\fP is replaced by
\fBXmATTACH_FORM\fP, and the child is attached to the bottom side of the
Form\&.
.IP "\fBXmATTACH_OPPOSITE_WIDGET\fP" 10
Attach the bottom side of the child to
the bottom side of the widget or
gadget specified in the \fBXmNbottomWidget\fP resource\&.
.IP "\fBXmATTACH_POSITION\fP" 10
Attach the bottom side of the child to a
position that is relative to
the top side of the Form and in proportion to the height of the Form\&.
This position is determined by the \fBXmNbottomPosition\fP and
\fBXmNfractionBase\fP resources\&.
.IP "\fBXmATTACH_SELF\fP" 10
Attach the bottom side of the child to a position
that is proportional
to the current \fIy\fP value of the bottom of the child divided by the
height of the Form\&.
This position is determined by the \fBXmNbottomPosition\fP and
\fBXmNfractionBase\fP resources\&.
\fBXmNbottomPosition\fP is set to a value proportional to the current
\fIy\fP value of the bottom of the child divided by the height of the
Form\&.
.RE
.IP "\fBXmNbottomOffset\fP" 10
Specifies the constant offset between the bottom side of the
child and the object to which it is
attached\&.
The relationship established remains, regardless of any resizing operations
that occur\&.
When this resource is explicitly set, the value of \fBXmNverticalSpacing\fP
is ignored\&.
.IP "\fBXmNbottomPosition\fP" 10
This resource is used to determine the position of the bottom side of
the child when the child\&'s \fBXmNbottomAttachment\fP is set to
\fBXmATTACH_POSITION\fP\&.
In this case the position of the bottom side of the child is relative to
the top side of the Form and is a fraction of the height of the Form\&.
This fraction is the value of the child\&'s \fBXmNbottomPosition\fP
resource divided by the value of the Form\&'s \fBXmNfractionBase\fP\&.
For example, if the child\&'s \fBXmNbottomPosition\fP is 50, the Form\&'s
\fBXmNfractionBase\fP is 100, and the Form\&'s height is 200, the position
of the bottom side of the child is 100\&.
.IP "\fBXmNbottomWidget\fP" 10
Specifies the widget or gadget to which
the bottom side of the child is attached\&.
This resource is used if the \fBXmNbottomAttachment\fP resource
is set to either \fBXmATTACH_WIDGET\fP
or \fBXmATTACH_OPPOSITE_WIDGET\fP\&.
.IP "" 10
A string-to-widget resource converter is automatically installed for use
with this resource\&.
With this converter, the widget that is to be the value of the resource
must exist at the time the widget that has the resource is created\&.
.IP "\fBXmNleftAttachment\fP" 10
Specifies attachment of the near (left) side of the child\&.
(Note that whether this resource actually applies to the left or
right side of the child and its attachment
depends on the value of
the \fBXmNlayoutDirection\fP resource\&.)
It can have the following values:
.RS
.IP "\fBXmATTACH_NONE\fP" 10
Do not attach the left side of the child\&.
If \fBXmNrightAttachment\fP is also \fBXmATTACH_NONE\fP, this value is
ignored and the child is given a default left attachment\&.
.IP "\fBXmATTACH_FORM\fP" 10
Attach the left side of the child to the left
side of the Form\&.
.IP "\fBXmATTACH_OPPOSITE_FORM\fP" 10
Attach the left side of the child to the
right side of the Form\&.
\fBXmNleftOffset\fP can be used to determine the visibility of the
child\&.
.IP "\fBXmATTACH_WIDGET\fP" 10
Attach the left side of the child to the right
side of the widget or
gadget specified in the \fBXmNleftWidget\fP resource\&.
If \fBXmNleftWidget\fP is NULL, \fBXmATTACH_WIDGET\fP is replaced by
\fBXmATTACH_FORM\fP, and the child is attached to the left side of the
Form\&.
.IP "\fBXmATTACH_OPPOSITE_WIDGET\fP" 10
Attach the left side of the child to
the left side of the widget or
gadget specified in the \fBXmNleftWidget\fP resource\&.
.IP "\fBXmATTACH_POSITION\fP" 10
Attach the left side of the child to a
position that is relative to
the left side of the Form and in proportion to the width of the Form\&.
This position is determined by the \fBXmNleftPosition\fP and
\fBXmNfractionBase\fP resources\&.
.IP "\fBXmATTACH_SELF\fP" 10
Attach the left side of the child to a position
that is proportional to
the current \fIx\fP value of the left side of the child divided by the
width of the Form\&.
This position is determined by the \fBXmNleftPosition\fP and
\fBXmNfractionBase\fP resources\&.
\fBXmNleftPosition\fP is set to a value proportional to the current
\fIx\fP value of the left side of the child divided by the width of the
Form\&.
.RE
.IP "\fBXmNleftOffset\fP" 10
Specifies the constant offset between the near (left) side of the
child and the object to which it is attached\&.
(Note that whether this resource actually applies to the left or
right side of the child and its attachment
depends on the value of
the \fBXmNlayoutDirection\fP resource\&.)
The relationship established remains, regardless of any resizing operations
that occur\&.
When this resource is explicitly set, the value of \fBXmNhorizontalSpacing\fP
is ignored\&.
.IP "\fBXmNleftPosition\fP" 10
This resource is used to determine the position of the near (left) side
of the child when the child\&'s \fBXmNleftAttachment\fP is set to
\fBXmATTACH_POSITION\fP\&.
(Note that whether this resource actually applies to the left or
right side of the child and its attachment
depends on the value of
the \fBXmNlayoutDirection\fP resource\&.)
.IP "" 10
In this case, the position of the left side of the child is relative to
the left side of the Form and is a fraction of the width of the Form\&.
This fraction is the value of the child\&'s \fBXmNleftPosition\fP resource
divided by the value of the Form\&'s \fBXmNfractionBase\fP\&.
For example, if the child\&'s \fBXmNleftPosition\fP is 50, the Form\&'s
\fBXmNfractionBase\fP is 100, and the Form\&'s width is 200, the position
of the left side of the child is 100\&.
.IP "\fBXmNleftWidget\fP" 10
Specifies the widget or gadget to which the near (left) side of the
child is attached\&.
(Note that whether this resource actually applies to the left or
right side of the child and its attachment
depends on the value of
the \fBXmNlayoutDirection\fP resource\&.)
The \fBXmNleftWidget\fP resource is used if the \fBXmNleftAttachment\fP
resource is set to either \fBXmATTACH_WIDGET\fP
or \fBXmATTACH_OPPOSITE_WIDGET\fP\&.
.IP "" 10
A string-to-widget resource converter is automatically installed for use

ChangeLog
2017-03-28 01:32:30 +0300 Oleksiy Chernyavskyy 

	* removed redundant includes (HEAD, origin/master, origin/HEAD, master)

2017-03-28 01:11:59 +0300 Oleksiy Chernyavskyy 

	* removed bugfix 1565. See details on http://bugs.motifzone.net/show_bug.cgi?id=1565

2016-11-09 07:19:53 +0200 Alex 

	* obsolete malloc.h changed on stdlib.h

2016-06-11 05:03:21 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-11 05:02:55 +0300 Oleksiy Chernyavskyy 

	* updated release 2.3.6 information

2016-06-11 04:57:54 +0300 Oleksiy Chernyavskyy 

	* fixid parallel building (make -j) error

2016-06-11 04:57:20 +0300 Oleksiy Chernyavskyy 

	* removed automatic configure start from autogen.sh

2016-06-11 04:26:48 +0300 Oleksiy Chernyavskyy 

	* minor fix

2016-06-11 04:21:53 +0300 Oleksiy Chernyavskyy 

	* removed extra dist Imakefile from all Makefile.am files

2016-06-11 04:06:37 +0300 Oleksiy Chernyavskyy 

	* Purified copyright headers (Motif is LGPL covered). Removed many Imakefile's and Imake-related stuff

2016-06-10 23:40:43 +0300 Oleksiy Chernyavskyy 

	* updated release notes and version 2.3.5 -> 2.3.6

2016-06-10 23:27:56 +0300 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-06-10 23:15:00 +0300 Oleksiy Chernyavskyy 

	* bug 1654 fix

2016-06-10 21:58:16 +0300 Oleksiy Chernyavskyy 

	* old patches. need review

2016-03-29 03:48:59 +0300 Oleksiy Chernyavskyy 

	* removed obsolete files

2016-03-17 04:23:17 +0200 Oleksiy Chernyavskyy 

	* release 2.3.5 (tag: release_2_3_5)

2016-03-16 23:07:12 +0200 Oleksiy Chernyavskyy 

	* updated ChangeLog

2016-03-16 23:06:24 +0200 Oleksiy Chernyavskyy 

	* added mkchangelog script for automated generation of the ChangeLog

2016-03-16 00:46:49 +0200 Oleksiy Chernyavskyy 

	* Reimplemented bugfix 1565

2016-02-26 03:36:51 +0200 Oleksiy Chernyavskyy 

	* Updated RPM creation instructions

2014-08-20 15:45:41 +0300 Oleksiy Chernyavskyy 

	* fixed bug 1636

2013-08-21 11:49:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1612 (Label size computed wrong within a Form).

2013-07-11 18:14:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1618 (Eliminates 29 warnings of some type in modules from "demos" catalog).

2013-07-11 15:20:54 +0300 Mykola Vshyvkov 

	* Fixed bug #1616 (Eliminates 23 warnings "suggest parentheses around assignment used" in modules from "demos" catalog).

2013-07-11 15:01:50 +0300 Mykola Vshyvkov 

	* Fixed bug #1617 (Eliminates 23 warnings "cast to pointer from integer of different size" in modules from "demos" catalog).

2013-07-11 12:35:30 +0300 Mykola Vshyvkov 

	* Fixed bug #1615 (Eliminates 35 warnings "enumeration value XXX not handled" in modules from "demos" catalog).

2013-07-10 06:55:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1614 (Eliminates 36 warnings "variable may be used uninitialized" in modules from "demos" catalog).

2013-07-10 05:05:28 +0300 Mykola Vshyvkov 

	* Fixed bug #1613 (Eliminates 60 warnings "implicit declaration of function" in modules from "demos" catalog).

2013-07-04 14:08:09 +0300 Mykola Vshyvkov 

	* Fixed bug #1611 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls in modules of "clients" catalog).

2013-07-03 21:27:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1610 (Eliminates 15 warnings of some types in modules of "clients" catalog).

2013-07-03 18:52:21 +0300 Mykola Vshyvkov 

	* Fixed bug #1609 (Eliminates 18 warnings  of "incompatible implicit declaration" in modules of "clients" catalog).

2013-07-03 14:58:22 +0300 Mykola Vshyvkov 

	* Fixed bug #1608 (Eliminates 30 warnings  of "array subscript has type 'char'" in modules of "clients" catalog).

2013-07-03 14:35:04 +0300 Mykola Vshyvkov 

	* Fixed bug #1607 (Eliminates 92 warnings  of "variable may be used uninitialized" in modules of "clients" catalog).

2013-07-03 12:48:55 +0300 Mykola Vshyvkov 

	* Fixed bug #1606 (Eliminates 130 warnings  of "expects argument of type" in modules of "clients" catalog).

2013-07-02 13:40:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 20:52:37 +0300 Mykola Vshyvkov 

	* Fix of Bug #1604 fix.

2013-07-01 19:07:32 +0300 Mykola Vshyvkov 

	* Fixed bug #1604 (Was been replaced  depecated code with using XKeycodeToKeysym Xlib function calls).

2013-07-01 19:02:25 +0300 Mykola Vshyvkov 

	* Fixed bug #1605 (Eliminates 14 warnings of some types in WML tool).

2013-06-26 23:45:14 +0300 Mykola Vshyvkov 

	* Fixed bug #1603 (Eliminates 18 warnings of some types in some modules).

2013-06-26 19:21:20 +0300 Mykola Vshyvkov 

	* Fixed bug #1602 (Eliminates 6 warnings "implicit declaration of function" in some modules).

2013-06-26 18:36:47 +0300 Mykola Vshyvkov 

	* Fixed bug #1601 (Eliminates 6 warnings: "suggest explicit braces to avoid ambiguous" in some modules).

2013-06-26 17:40:00 +0300 Mykola Vshyvkov 

	* Fixed bug #1600 (Eliminates 26 warnings about using uninitialized variables in some modules).

2013-06-26 15:34:53 +0300 Mykola Vshyvkov 

	* Fixed bug #1599 (Eliminates 14 warnings of incompatible type argument from some modules).

2013-06-25 17:37:02 +0300 Mykola Vshyvkov 

	* Updated ChangeLog.

2013-06-25 13:57:31 +0300 Mykola Vshyvkov 

	* Merge branch 'openmotif_2_3' of ssh://git.code.sf.net/p/motif/code into openmotif_2_3

2013-06-25 12:55:13 +0300 Nikolay Vs 

	* Fixed bug #1595. (Patched Configure.ac). Patch by Ulrich Muller.

2013-06-25 12:55:13 +0300 Nikolay Vs 

	* Added FIX #1595. Thanks to Ulrich Muller

2013-03-29 17:37:29 +0200 Yuriy Syrota 

	* Fixed bug #1593 (Missing fontconfig lib in linking, due to outdated test in ac_find_xft.m4). Patch by Ulrich Müller.

2013-03-28 17:48:58 +0200 Yuriy Syrota 

	* Fixed bug #1594 (configure: typo in warning "No splint is dound"). Thanks to Ulrich Müller.

2012-10-31 15:59:57 +0200 Yuriy Syrota 

	* Fixed bug #1581 (Buffer overrun in FontS.c)

2012-10-22 15:45:31 +0300 Yuriy Syrota 

	* Updated link to documentation (tag: release_2_3_4)

2012-10-16 16:47:45 +0300 Yuriy Syrota 

	* Updated Release Notes


demos/unsupported/xmform/xmform.c
/* $XConsortium: xmform.c /main/4 1995/07/15 20:47:50 drk $ */
/*
 * Motif
 *
 * Copyright (c) 1987-2012, The Open Group. All rights reserved.
 *
 * These libraries and programs are free software; you can
 * redistribute them and/or modify them under the terms of the GNU
 * Lesser General Public License as published by the Free Software
 * Foundation; either version 2 of the License, or (at your option)
 * any later version.
 *
 * These libraries and programs are distributed in the hope that
 * they will be useful, but WITHOUT ANY WARRANTY; without even the
 * implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR
 * PURPOSE. See the GNU Lesser General Public License for more
 * details.
 *
 * You should have received a copy of the GNU Lesser General Public
 * License along with these librararies and programs; if not, write
 * to the Free Software Foundation, Inc., 51 Franklin Street, Fifth
 * Floor, Boston, MA 02110-1301 USA
 */
/*
 * HISTORY
 */
/**---------------------------------------------------------------------
***
***	file:		xmform.c
***
***	project:	Motif Widgets example programs
***
***	creation:	Spring 91 - Cambridge
***                       Daniel Dardailler.
***
***	description:	This program demonstrates the possibilities of
***                     the Motif form widget.
***                     It can display a string of letters made from buttons
***	                and arrows (resizable using Form attachments).
***
***	defaults:	xmform looks nicer with these defaults:
***
xmform*highlightThickness:       0
xmform.XmForm.shadowThickness:   3
xmform*XmPushButton.background:  cadetblue
xmform*XmArrowButton.foreground: cadetblue
xmform*XmArrowButton.background: lightgrey
xmform*XmForm.background:        lightgrey
xmform*topShadowColor:           white
xmform*bottomShadowColor:        black
***-------------------------------------------------------------------*/

#include <stdlib.h>

#include <Xm/Xm.h>
#include <Xm/Form.h>
#include <Xm/PushB.h>
#include <Xm/ArrowB.h>

static void FillItPlease();
static void NearlyEvenSpread();
static Widget AddStem() ;

int main(argc, argv)
int argc; char **argv;
{
    XtAppContext app_context;
    Widget      toplevel, topform, *forms ;
    Cardinal len_word, p ;
    String word = "MOTIF" ;
    char s[2] ;

    toplevel = XtAppInitialize(&app_context, "XMdemos", NULL, 0,
			       &argc, argv, NULL, NULL, 0);
    topform = XmCreateForm(toplevel, "topform", NULL, 0);
    XtManageChild(topform);

    /* pick up the string to be displayed if any */
    if (argc == 2) word = argv[1] ;

    /* create a form for each letter and fill it with the 'stem widgets' */
    len_word = strlen(word);
    forms = (Widget*)calloc(len_word,sizeof(Widget));
    s[1] = 0 ;  /* string used for the subform widget name */
    for (p = 0 ; p < len_word; p++) {
	s[0] = word[p] ;  /* subform  names are the letters themselves */
	forms[p] = XmCreateForm(topform, s, NULL, 0);
	FillItPlease(forms[p], word[p]);
    }

    /* evenly spread out the forms with some spacing in between */
    NearlyEvenSpread(forms, p,
		     2,          /* horizontal spacing */
		     9) ;        /* vertical margin */

    XtManageChildren(forms, p);
    XtRealizeWidget(toplevel);
    XtAppMainLoop(app_context);

    return 0;    /* make compiler happy */
}


static void NearlyEvenSpread(child_list, num_child,
			     hor_spacing, ver_spacing)
Widget * child_list;
Cardinal num_child;
Dimension hor_spacing;
Dimension ver_spacing;
{
    /* this algorithm works only for the default fractionBase == 100 */
    /* hor_spacing and ver_margin are equally applied on each sides */

    Dimension hor_child_room ;
    Cardinal n, i ;
    Arg args[10] ;

    hor_child_room = 100 / num_child ;

    for (i = 0 ; i < num_child ; i++) {
	n = 0 ;
	XtSetArg(args[n], XmNleftAttachment, XmATTACH_POSITION); n++ ;
	XtSetArg(args[n], XmNleftPosition,
		 hor_child_room*i + hor_spacing); n++ ;
	XtSetArg(args[n], XmNrightAttachment, XmATTACH_POSITION); n++ ;
	XtSetArg(args[n], XmNrightPosition,
		 hor_child_room*(i+1) - hor_spacing); n++ ;
	XtSetArg(args[n], XmNtopAttachment, XmATTACH_POSITION); n++ ;
	XtSetArg(args[n], XmNtopPosition, ver_spacing); n++ ;
	XtSetArg(args[n], XmNbottomAttachment, XmATTACH_POSITION); n++ ;
	XtSetArg(args[n], XmNbottomPosition, 100 - ver_spacing); n++ ;
	XtSetValues(child_list[i], args, n);
    }
}

static void FillItPlease(form, letter)
Widget form ;
char letter ;
{
    /* This routine only handle the "MOTIF" letters, it's up
       to you to implement the rest, good luck for K,N,R,V,X,Y,Z... */

    Cardinal n = 0, p = 0 ;
    Arg args[10] ;
    Widget child[10] ;

    switch(letter) {
    case 'M' :
	/* 2 buttons and the arrow in between on top */
	child[p] = AddStem(form, 0, 25, 100, 0, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 0, 100, 100, 75, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 0, 75, 35, 25, NULL); p++;
/*	XtSetArg(args[n], XmNshadowThickness, 0); n++;*/
	XtSetArg(args[n], XmNarrowDirection, XmARROW_DOWN); n++;
	XtSetValues(child[2], args, n);
	XtManageChildren(child, p);
	break ;
    case 'O' :
	/* 2 left and right vertical buttons and 2 horizontals */
	child[p] = AddStem(form, 25, 25, 75, 0, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 0, 100, 25, 0, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 25, 100, 75, 75, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 75, 100, 100, 0, XmPUSHBUTTON); p++;
	XtManageChildren(child, p);
	break ;
    case 'T' :
	/* 1 vertical button and 1 horizontal on top */
	child[p] = AddStem(form, 25, 63, 100, 38, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 0, 100, 25, 0, XmPUSHBUTTON); p++;
	XtManageChildren(child, p);
	break ;
    case 'I' :
	/* 1 vertical button + one arrow on top */
	child[p] = AddStem(form, 0, 63, 25, 38, NULL); p++;
	child[p] = AddStem(form, 30, 63, 100, 38, XmPUSHBUTTON); p++;
/*	XtSetArg(args[n], XmNshadowThickness, 0); n++;*/
	XtSetValues(child[0], args, n);
	XtManageChildren(child, p);
	break ;
    case 'F' :
	/* 1 vertical button and 2 horizontal on the right */
	child[p] = AddStem(form, 0, 25, 100, 0, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 0, 100, 25, 25, XmPUSHBUTTON); p++;
	child[p] = AddStem(form, 36, 75, 61, 25, XmPUSHBUTTON); p++;
	XtManageChildren(child, p);
	break ;
    default:
	child[p] = AddStem(form, 0, 100, 100, 0, XmPUSHBUTTON); p++;
	XtManageChildren(child, p);
	break ;
    }
}


static Widget AddStem(form, top, right, bottom, left, type)
Widget form ;
Dimension top, right, bottom, left;
int type ;
{
    /* create an object, either a button or an arrow, using
       positionnal attachment everywhere, instead of possibly
       form attachment or widget attachment */

    Cardinal n = 0 ;
    Arg args[10] ;

    XtSetArg(args[n], XmNtopAttachment, XmATTACH_POSITION); n++;
    XtSetArg(args[n], XmNtopPosition, top); n++;
    XtSetArg(args[n], XmNrightAttachment, XmATTACH_POSITION); n++;
    XtSetArg(args[n], XmNrightPosition, right); n++;
    XtSetArg(args[n], XmNbottomAttachment, XmATTACH_POSITION); n++;
    XtSetArg(args[n], XmNbottomPosition, bottom); n++;
    XtSetArg(args[n], XmNleftAttachment, XmATTACH_POSITION); n++;
    XtSetArg(args[n], XmNleftPosition, left); n++;
    if (type == XmPUSHBUTTON)
	return XmCreatePushButton(form, "", args, n);
    else
	return XmCreateArrowButton(form, "", args, n);
}
